walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,159 @@
+
+    // stores the index in the array from which this meeting begins
+    int startPoint = 0;
+    // stores the time at which this meeting begins
+    int startTime = 0;
+    // stores the index in the array at which this meeting ends
+    int endPoint = 0;
+    // stores the time at which this meeting ends
+    int endTime = 0;
+
+    if (prev < 0) {
+      // If the meeting is to be inserted at the start of the list, startPoint = 0
+      startPoint = startPos;
+      startTime = meetingTime.start();
+    } else {
+      TimeRange prevElem = busyTimes.get(prev);
+      /*
+       * Otherwise find out whether this meeting overlaps with the previous meeting
+       * If it does, coalesce it with the previous meeting otherwise keep them
+       * separate
+       */
+      startPoint = prevElem.overlaps(meetingTime) ? prev : startPos;
+      startTime = prevElem.overlaps(meetingTime) ? prevElem.start() : meetingTime.start();
+    }
+
+    if (next >= busyTimes.size()) {
+      // If this element is to be inserted at the end of the list, endPoint = last element of list
+      endPoint = next - 1;
+      endTime = meetingTime.end();
+    } else {
+      /*
+       * Otherwise find out whether this meeting overlaps with the next meeting
+       * If it does, coalesce it with the next meeting otherwise keep them
+       * separate
+       */
+      TimeRange nextElem = busyTimes.get(next);
+      endPoint = nextElem.overlaps(meetingTime) ? next : next - 1;
+      endTime = nextElem.overlaps(meetingTime) ? nextElem.end() : meetingTime.end();
+    }
+
+    // create a new meeting with the given start and end times
+    TimeRange newBusy = TimeRange.fromStartEnd(startTime, endTime, false);
+
+    /*
+     * This meeting ends before the current first meeting starts, so add it to the
+     * beginning of the list
+     */
+    if (endPoint < 0) {
+      newBusyTimes.add(newBusy);
+    }
+    for (int i = 0; i < busyTimes.size(); i++) {
+      if (i == startPoint && endPoint >= 0 && startPoint < busyTimes.size()) {
+        // Replace all the entries between startPoint and endPoint with the coalesced entry
+        newBusyTimes.add(newBusy);
+        i = endPoint;
+      } else {
+        newBusyTimes.add(busyTimes.get(i));
+      }
+    }
+    /*
+     * This meeting starts after the current last meeting ends, so add it to the end of
+     * the list
+     */
+    if (startPoint >= busyTimes.size() && endPoint >= 0) {
+      newBusyTimes.add(newBusy);
+    }
+    return newBusyTimes;
+  }
+
+  /*
+   * Given a list of busy times in non-overlapping, sorted order and a meeting duration
+   * returns a list of time slots of atleast 'duration' length during which no meetings
+   * are scheduled
+   */
+  private ArrayList<TimeRange> findFreeTimes(ArrayList<TimeRange> busyTimes, long duration) {
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    ArrayList<TimeRange> freeTimes = new ArrayList<>();
+    int start = TimeRange.START_OF_DAY;
```

Comment:
Document what the `start` and `end` variable is. Also, do you need the `end` variable in the for loop?